### ISSUE TITLE
Materialized view: don't do read-before-write when view key columns are same as base's 

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#20679]
This issue is a re-focused version of an old issue https://github.com/scylladb/scylladb/issues/3594 that we recently closed - perhaps prematurely.

The need to read the old base row before updating a materialized view (so-called "read-before-write") in some cases dominates the performance of writes to a table with a materialized view, so if we can avoid this read-before-write, we should.

This issue is about the case when **the view's key has the same columns as the base's key** (in other words, there isn't a base-regular column in the view's key). The claim is that in this case, a read-before-write is NOT needed, because:

1. In this case, an update to one row in the base is translated to an update of one row in the view - there's no deletion of old row or need to re-write the entire row.
2. The _key_ of this view row update is fully known and doesn't need to be read from the table: The view key is composed of the same key columns of the base table, so they are known from the base-table update (you can't have an update that doesn't fully specify the key).
4. If some base-table columns are not present in the base update, it is fine to also send not send them in the view update - the old value will remain in the view, just like it remained in the base. There is no need to read the entire base row and write it again.
5. I believe (but it should be verified) that there is no need to read the entire row just for calculating row marker timestamps - I think that mess was replaced by virtual cells in https://github.com/scylladb/scylladb/issues/3362.

Note that a read-before-write IS needed in the other case - when there is regular column (call it R) in the base that is part of the view's key:
1. If the update doesn't change R, we need to read it to know which view row to modify.
2. If the update does include R, we don't know whether it's the same R value as was previously saved... So we need to read R's old value.

But even in this case, there is a (probably not very exciting) case where read-before-write can be avoided: When the update does not contain a mutation of R or of any other selected column in the view. Remember that in the case of regular-base-column-is-in-view-key, we _do not_ have virtual columns, so modifying an unselected column does not necessitate modifying the view at all, and there's also no need to read-before-write when we aren't going to write to the view.

Don't forget that when a base table has multiple views, we can only avoid the read-before-write if all the views don't need it.

CC @mlitvk @vladzcloudius

**Comment by nyh** [comment]
@vladzcloudius rightly pointed out that the same optimization should also apply to a secondary index of a column that was already one of the base table's key columns. The secondary index's view does contain an extra key column - the "token" column - but it's a computed function of the base-table key columns so the read-before-write should still not be needed.

It's interesting to note that although the secondary-index view in the above assumption only selects key columns so ideally view rows should never even be changed - only created and deleted - today this is **not** the case because today in this case the index's view has so-called virtual columns, so every update to any column in the base table does cause an update to the view (writing empty cells with the timestamps from the base update). Issue https://github.com/scylladb/scylladb/issues/3833 is about avoiding the need for virtual columns in the SI case, but it's orthogonal to the issue discussed here because even with the virtual columns there is no need for read-before-write.

**Comment by nyh** [comment]
The following easy patch avoids the unnecessary read-before-write:
```patch
--- a/replica/table.cc
+++ b/replica/table.cc
@@ -3477,6 +3477,17 @@ future<row_locker::lock_holder> table::do_push_view_replica_updates(shared_ptr<d
     if (views.empty()) {
         co_return row_locker::lock_holder();
     }
+    // If we do have views but all of them have the same key columns as the
+    // base (i.e., none has_base_non_pk_columns_in_view_pk), then the known
+    // base mutation can just be propagated to the views, and we don't need
+    // a read before write (issue #20679).
+    if (std::ranges::none_of(views, [] (const auto& v) {
+            return v.view->view_info()->has_base_non_pk_columns_in_view_pk(); })) {
+        co_await gen->generate_and_propagate_view_updates(*this, base, sem.make_tracking_only_permit(s, "push-view-updates-3", timeout, tr_state), std::move(views), std::move(m), { }, tr_state, now, timeout);
+        // No read-before-write, so no lock is needed.
+        co_return row_locker::lock_holder();
+    }
+
     auto cr_ranges = co_await db::view::calculate_affected_clustering_ranges(gen->get_db().as_data_dictionary(), *base, m.decorated_key(), m.partition(), views);
```

But unfortunately, more fixes will be needed to fix incorrect assumptions that other code made. For example, `view_updates::update_entry()` assumes that if the update mutation deletes a certain column and it's also unset in "existing", we don't need to do anything. But now that we don't actually read "existing", we **do** need to apply the deletion in the view because we don't need if it's superflous or not. The update_entry() code - and maybe other code, should need that "existing" wasn't read at all and not try to avoid work by comparing it to existing.

**Comment by nyh** [comment]
By the way, @vladzcloudius if you want to very quickly measure the **performance** of disabling read-before-write, you can apply the above patch. It generates wrong results in some edge cases (which is of course must be fixed) but the performance should be the same so can already be measured.

**Comment by mykaul** [comment]
> But unfortunately, more fixes will be needed to fix incorrect assumptions that other code made. For example, `view_updates::update_entry()` assumes that if the update mutation deletes a certain column and it's also unset in "existing", we don't need to do anything. But now that we don't actually read "existing", we **do** need to apply the deletion in the view because we don't need if it's superflous or not. The update_entry() code - and maybe other code, should need that "existing" wasn't read at all and not try to avoid work by comparing it to existing. Right now with the above patch, the cql-pytest `test_materialized_view.py::test_mv_with_only_primary_key_rows` fails.

Do we have tests catching the above? I'm more and more concerned that we have such issues - and either do not have tests to cover them and do not have a tool to check 'correctness' of MVs.

**Comment by nyh** [comment]
> Do we have tests catching the above? I'm more and more concerned that we have such issues - and either do not have tests to cover them and do not have a tool to check 'correctness' of MVs.

Yes, the first bug of just applying the above patch I caught was caught by a cql-pytest. Actually most of the MV tests are in Boost tests, but I haven't tested those yet, it's possible they will catch additional bugs.

I believe our coverage of MV correctness is pretty good - over the years (not just recently) every correctness issue we fixed or even considered was covered in those tests. We now have 7000 lines of test in Boost, 3000 lines in cql-pytest and 4000 lines of dtest. So it will be hard to break MV without triggering on those tests. But I can't say with 100% certainty that every bug will be detected, or how much "code coverage" these tests had (Eliran's coverage project would have come in handy now).

The bad news is that it turns out that we have A LOT of code which seems to do all sort of logic and optimizations based on the "existing" row, the "update row", and the bazillion ways they can interact, and now I need to understand how all that hard to understand and super-fragile code interacts with the case where we **deliberately did not read "existing"** - which is not the same as an empty existing. This code mutated frequently over the years - every patch for breaking up large changes to chunks, patch for adding indexing of static columns, collections, and a lot of other patches - touched it. The code is now super convoluted and somewhat duplication-heavy, and I'm having a difficult time reviewing it. Also frustratingly, I keep finding pieces of code which look like "obvious" bugs, but when I try to reproduce a bug I see that it doesn't - the code looks wrong but works properly ;-)

**Comment by nyh** [comment]
> A small comment about the patch, @nyh: you can cache the fact that PK columns of the MV and its base are the same instead of calculating for very write.

I don't know if it's possible - note that even the list of views (`views`) is calculated right now on-the-fly because different writes may affect different views (e.g., an update to an unselected column might does not affect a view which has a new key column) - so the boolean flag of whether "all relevant views don't have a new key column" may be different every time. That being said, I'm not sure.

As they say, I still have bigger fish to fry :-( The code is super convoluted and uses the "existing" row (from read-before-write) in many different places and I'm now trying to debug failing tests (which luckily I have). I'm no longer optimistic that I'll finish this today - although as I said the patch above is all you need if you want to measure performance and don't need (yet) correctness :-)

**Comment by nyh** [comment]
The state in the view is already a variable. But the need for read-before-write depends not on one view but on the views that are relevant or not for this specific write and may depend on what this write contains. I just figured out in my debugging that, for example, a range tombstone (spanning multiple rows) requires a read before write.

**Comment by vladzcloudius** [comment]
> The state in the view is already a variable. But the need for read-before-write depends not on one view but on the views that are relevant or not for this specific write and may depend on what this write contains. 

You are right! I missed that - I retract my previous comment then. ;)

> I just figured out in my debugging that, for example, a range tombstone (spanning multiple rows) requires a read before write.

In some cases, yes (depending on the actual structure of the MV clustering portion of the primary key). It's an acceptable "limitation" IMO.

**Comment by nyh** [comment]
Sent a pull request: https://github.com/scylladb/scylladb/pull/20984

**Comment by nyh** [comment]
A failed test in the pull request https://github.com/scylladb/scylladb/pull/20984 found a fundamental flaw with the original idea raised in this issue, of avoiding read-before-write:

The problem is that although (in our case) base rows and view rows have a one-to-one correspondence, the same can't be said about partition tombstones:

* A single partition tombstone (or range tombstone) may span many view rows, and without read-before-write, we don't know which rows.
* This is easy to solve when the partition tombstone is written *after* the data: Simply, for a partition tombstone **don't** avoid a read-before-write. The read-before-write of the existing partition allows us to iterate over its rows and write tombstones to the corresponding view rows and apply the tombstone to all of them. This is what I did in the pull request, and it works well.
* But the big problem is when the writes come in opposite order: First the partition tombstone and then a single row with an _older_ timestamp. How do we know when we write that single row at the end, that we need to do a read-before-write to discover the partition tombstone? 

In some sense, this problem kills the entire idea proposed in this issue. Can we save it somehow?

I thought of two different ways to "save" this idea, but they come with big restrictions that we need to discuss if users could live with:

1. We can invent a flag that forbids writing partition (or range) tombstones to a table with a materialized views. On such a table, the problem described above can't happen. Note that DynamoDB users are fine with not having a partition deletion feature (you can only delete individual rows), so maybe our users could live with this restriction too? Another point worth noting is that we recently had multiple performance problems with partition deletions and MV (taking too much memory, as well as unavoidable timeouts on the very long operation) - so that's another reason to forbid such operation.

2. Another option is to do read-before-write and not avoid it - but, with a twist: Only read the partition tombstone and not the row itself. Since we assume long partitions, there will only be a few of those partition tombstones and we'll be reading the same ones over and over, so they can frequently come from cache - as opposed to reading each individual row which is usually not cached. Note that one of the _worst cases_ for the read-before-write problem is when writing to a new row (that didn't yet exist) in a long partition; In that case, the row doesn't exist in cache to we need to access disk just to find it doesn't exist there too (the Bloom filter is per-partition, so it doesn't help know that a row doesn't exist in a long partition) - so we have uncached reads for every read-before-write of the row. In contrast, if we only need to read-before-write the partition tombstone, it will most likely be in cache. **A problem:** I'm not sure if or how this solution works for range tombstones, if it works at all :-( Maybe we'll be able to allow partition tombstones in this case, but not range tombstones?

I think solution 1 (the "no partition/range deletion"), as unfortunate at as it is, is the only way to save this optimization idea. But maybe someone else can come up with a better solution?

**Comment by mykaul** [comment]
@tgrabiec - can you answer the above questions on range tombstones in cache (see item 2 in the comment above) ?

**Comment by kostja** [comment]
* nyh ***@***.***> [24/10/09 09:52]:
> 1. We can invent a flag that forbids writing partition (or range) tombstones to a table with a materialized views. On such a table, the problem described above can't happen. Note that DynamoDB users are fine with not having a partition deletion feature (you can only delete individual rows), so maybe our users could live with this restriction too? Another point worth noting is that we recently had multiple performance problems with partition deletions and MV (taking too much memory, as well as unavoidable timeouts on the very long operation) - so that's another reason to forbid such operation.
> 

We can combine a base table write into the base table with a special
tombstone that, when collapsed with apartition tombstone, converts
into a row tombstone, but when collapsed with a live row,
disappears.

-- 
Konstantin Osipov, Moscow, Russia

**Comment by tgrabiec** [comment]
> [@tgrabiec](https://github.com/tgrabiec) - can you answer the above questions on range tombstones in cache (see item 2 in the comment above) ?

I didn't see any question about cache, just statements. What's the question?

**Comment by tgrabiec** [comment]
> I'm not sure if or how this solution works for range tombstones, if it works at all :-( Maybe we'll be able to allow partition tombstones in this case, but not range tombstones?

There is no way to read just range tombstones without also reading the rows from cache (or any other mutation source: memtables, sstables).

**Comment by nyh** [comment]
> I don't understand something. If we write view updates using the same timestamp as the base table update, why can't we rely on mutation merging to do the right thing when applying tombstones? When you apply a partition tombstone to the base table, you also apply it to the view, and it will delete exactly the same rows in the view which are deleted in the base table. Since rows in the view have the same timestamps as rows in the base table.

You are forgetting that although the base and view may have the one-to-one correspondence between _rows_ (this is the case addressed in this patch, where the primary key of base and view has the same columns), the same is not true for **partitions**: Consider that the base has primary key (p, c) and the view has primary key (c, p). Now a **partition** tombstone for p=1 in the base table is _not_ a partition tombstone in the view, we basically need to apply this tombstone to every view partition c, not to any specific partition. That's not something we can do... so the current code (after the fix for https://github.com/scylladb/scylladb/issues/5793) relies on read-before-write when updating view rows to check the tombstone in the _base_ partition.

As I mentioned above the "killer" scenario is when the the partition tombstone arrives **before** the earlier-timestamped data. This partition tombstone has no view rows to delete (there are none yet), and there is no place in the view to store it.

**Comment by tgrabiec** [comment]
> > I don't understand something. If we write view updates using the same timestamp as the base table update, why can't we rely on mutation merging to do the right thing when applying tombstones? When you apply a partition tombstone to the base table, you also apply it to the view, and it will delete exactly the same rows in the view which are deleted in the base table. Since rows in the view have the same timestamps as rows in the base table.
> 
> You are forgetting that although the base and view may have the one-to-one correspondence between _rows_ (this is the case addressed in this patch, where the primary key of base and view has the same columns), the same is not true for **partitions**: Consider that the base has primary key (p, c) and the view has primary key (c, p). Now a **partition** tombstone for p=1 in the base table is _not_ a partition tombstone in the view, we basically need to apply this tombstone to every view partition c, not to any specific partition. That's not something we can do... so the current code (after the fix for [#5793](https://github.com/scylladb/scylladb/issues/5793)) relies on read-before-write when updating view rows to check the tombstone in the _base_ partition.
> 
> As I mentioned above the "killer" scenario is when the the partition tombstone arrives **before** the earlier-timestamped data. This partition tombstone has no view rows to delete (there are none yet), and there is no place in the view to store it.

Right. I thought that the constraint is that primary key is the same between base and view, which is not the case.

**Comment by nyh** [comment]
> Right. I thought that the constraint is that primary key is the same between base and view, which is not the case.

Definitely not the entire primary key is the same between the base and the view - that would be a very boring and wasteful view ;-) 

But it is actually possible to have a view with the same **partition key** as the base table (these are called "LSI" in DynamoDB). For such cases, the bug I outlined here may not exist - we can write the same partition tombstone on the view (I didn't test it actually works today, but it can after we recently added exactly this optimization). But this further restriction makes this optimization even less common-place than it started, and as far as I understand (@vladzcloudius can you confirm), the user that was interested in this optimization doesn't have the "local" view (same partition key in base and view) use case.

**Comment by nyh** [comment]
Adding to what I wrote above: If the partition key is identical in base and view it solves (at least in theory) the problem of partition tombstones, but not of arbitrary clustering range tombstones... For those to work correctly too - and match between base and view -we indeed need all the clustering key columns to be the same too, so we end up requiring that the entire primary key is the same.

I called this case "very boring and wasteful view", but it may have some legitimate use cases such as filtering only some of the rows or removing some large columns (e.g., a column that typically contains a large string) to make scans more efficient. But this is become a very obscure and rare case for which this optimization is safe :-(
